sympy/ntheory/residue_ntheory.py

```diff
@@ -988,3 +989,150 @@ def eval(cls, n):
             if any(i > 1 for i in a.values()):
                 return S.Zero
             return S.NegativeOne**len(a)
+
+
+def _discrete_log_trial_mul(n, a, b, order=None):
+    a %= n
+    b %= n
+    if order is None:
+        order = n
+    x = 1
+    k = 1
+    for i in range(order):
+        if x == a:
+            return i
+        x = x * b % n
+    raise ValueError("Log does not exist")
+
+
+def _discrete_log_shanks_steps(n, a, b, order=None):
+    a %= n
+    b %= n
+    if order is None:
+        order = n_order(b, n)
+    m = isqrt(order) + 1
+    T = dict()
+    x = 1
+    for i in range(m):
+        T[x] = i
+        x = x * b % n
+    z = mod_inverse(b, n)
+    z = pow(z, m, n)
+    x = a
+    for i in range(m):
+        if x in T:
+            return i * m + T[x]
+        x = x * z % n
+    raise ValueError("Log does not exist")
+
+
+def _discrete_log_pollard_rho(n, a, b, order=None):
+    a %= n
+    b %= n
+
+    if order is None:
+        order = n_order(b, n)
+
+    for i in range(10):
```

Comment:
I'm assuming the 10 here is arbitrary. Maybe it should be `for i in range(retries)` where `retries=10` is a keyword argument.